source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    bool
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::is_recording() const
+    {
+      return is_recording_flag;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    types::tape_index
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::active_tape() const
+    {
+      return active_tape_index;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    bool
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::is_registered_tape(
+      const types::tape_index &tape_index) const
+    {
+      return registered_tapes.find(tape_index) != registered_tapes.end();
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::print(
+      std::ostream &stream) const
+    {
+      stream << "Active tape index: " << active_tape_index << "\n";
+      stream << "Registered tapes: ";
+      for (unsigned int i = 0; i < registered_tapes.size(); ++i)
+        stream << *(std::next(registered_tapes.begin(), i))
+               << (i < (registered_tapes.size() - 1) ? "," : "");
+      stream << "\n";
+      stream << "Recording? " << is_recording() << "\n";
+      stream << "Keep values? " << keep_values << "\n";
+      stream << "Use stored tape buffer sizes? "
+             << use_stored_taped_buffer_sizes << "\n";
+      stream << std::flush;
+
+      stream << "Touched independent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_independent_variables(); i++)
+        stream << touched_independent_variables[i]
+               << (i < (n_independent_variables() - 1) ? "," : "");
+      stream << std::endl;
+
+      stream << "Independent variable values: "
+             << "\n";
+      print_values(stream);
+
+      stream << "Touched marked independent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_independent_variables(); i++)
+        stream << touched_marked_independent_variables[i]
+               << (i < (n_independent_variables() - 1) ? "," : "")
+               << std::flush;
+      stream << std::endl;
+
+      stream << "Dependent variable values: "
+             << "\n";
+      for (unsigned int i = 0; i < n_dependent_variables(); i++)
+        stream << dependent_variables[i]
+               << (i < (n_dependent_variables() - 1) ? "," : "") << std::flush;
+      stream << std::endl;
+
+      stream << "Touched dependent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_dependent_variables(); i++)
+        stream << touched_dependent_variables[i]
+               << (i < (n_dependent_variables() - 1) ? "," : "") << std::flush;
```

Comment:
we really ought to invent a function that takes a container and outputs it with separators, rather than having to repeat this whole block over and over.

no need to do this here and now, but maybe keep it in mind for the future.

Reply:
Separately, the `std::flush` is really not very useful here -- you later push `std::endl` into the stream, which implies a flush. The only reason you may want to do the flush here is if you anticipate that any of the following operations triggers an `abort` and you're afraid that you won't make it to the `std::endl` -- though I can't really see how that would happen here.